File: pattern9/take_out_v4/src/Node.java
package pattern9.take_out_v4.src;

import java.util.Iterator;
import java.util.LinkedList;
import java.util.List;

public class Node implements Iterable<Node> {
    private LinkedList<Node> children = new LinkedList<>();
    private Node parent = null;

    public boolean removeFromParent() {
        if (parent != null) {
            return parent.removeChild(this);
        }
        return false;
    }

    public boolean removeChild(Node child) {
        if (children.remove(child)) {
            child.parent = null;
            return true;
        }
        return false;
    }

    public void addChild(Node child) {
        children.add(child);
        child.parent = this;
    }

    public boolean isLeaf() {
        if (children.size() == 0) {
            return true;
        }
        return false;
    }

    public List<Node> getChildren(){
        return this.children;
    }

    @Override
    public Iterator<Node> iterator() {
        return new NodeIterator(this);
    }

}

File: pattern16/visitor/src/node.py
from collections.abc import Iterator


class Node():
    def __init__(self) -> None:
        self.children: list = list()
        self.parent: Node = None

    def removeFromParent(self) -> bool:
        if self.parent is not None:
            return self.parent.removeChild(self)
        return False

    def removeChild(self, child: "Node") -> bool:
        self.children.remove(child)
        child.parent = None
        return True

    def addChild(self, child: "Node"):
        self.children.append(child)
        child.parent = self

    def isLeaf(self):
        if len(self.children) == 0:
            return True
        return False

    def getChildren(self) -> list:
        return self.children

    def __iter__(self) -> Iterator:
        yield self
        if self.isLeaf():
            return
        #对子节点进行遍历
        childNode: "Node"
        for childNode in self.getChildren():
            yield from childNode.__iter__()
